src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,596 @@
+                     *       each invocation of integrate() which might prove
+                     *       costly.
+                     */
+                    public boolean rightIntegrate(R r, Downstream<? super RR> downstream) {
+                        // The following logic is highly performance sensitive
+                        return (rightGreedy || rightProceed)
+                                && (rightIntegrator.integrate(rightState, r, downstream)
+                                || rightGreedy
+                                || (rightProceed = false));
+                    }
+                }
+
+                return new GathererImpl<T, State, RR>(
+                        State::new,
+                        (leftGreedy && rightGreedy)
+                                ? Integrator.<State, T, RR>ofGreedy(State::integrate)
+                                : Integrator.<State, T, RR>of(State::integrate),
+                        (leftCombiner == Gatherer.defaultCombiner()
+                        || rightCombiner == Gatherer.defaultCombiner())
+                                ? Gatherer.defaultCombiner()
+                                : State::joinLeft,
+                        (leftFinisher == Gatherer.<A, R>defaultFinisher()
+                        && rightFinisher == Gatherer.<AA, RR>defaultFinisher())
+                                ? Gatherer.defaultFinisher()
+                                : State::finish
+                );
+            }
+        }
+    }
+
+    // Public built-in Gatherers and factory methods for them
+
+    /**
+     * Gathers elements into fixed-size windows. The last window may contain
+     * fewer elements than the supplied window size.
+     *
+     * <p>Example:
+     * {@snippet lang = java:
+     * // will contain: [[1, 2, 3], [4, 5, 6], [7, 8]]
+     * List<List<Integer>> windows =
+     *     Stream.of(1,2,3,4,5,6,7,8).gather(Gatherers.windowFixed(3)).toList();
+     * }
+     *
+     * @param windowSize the size of the windows
+     * @param <TR> the type of elements the returned gatherer consumes
+     *             and the contents of the windows it produces
+     * @return a new gatherer which groups elements into fixed-size windows
+     * @throws IllegalArgumentException when groupSize is less than 1
+     */
+    public static <TR> Gatherer<TR, ?, List<TR>> windowFixed(int windowSize) {
+        if (windowSize < 1)
+            throw new IllegalArgumentException("'groupSize' must be greater than zero");
+
+        return Gatherer.ofSequential(
+                // Initializer
+                () -> new ArrayList<>(windowSize),
+
+                // Integrator
+                Integrator.<ArrayList<TR>,TR,List<TR>>ofGreedy((window, e, downstream) -> {
+                    window.add(e);
+                    if (window.size() < windowSize) {
+                        return true;
+                    } else {
+                        var full = List.copyOf(window);
```

Comment:
Here, we are rejecting `null` stream elements for no particular reason. Let's use `SharedSecrets.getJavaUtilCollectionAccess().listFromTrustedArrayNullsAllowed(window.toArray())` instead?